src/IJob.php

```diff
@@ -3,11 +3,12 @@
 namespace Contributte\Scheduler;
 
 use DateTime;
+use DateTimeInterface;
 
 interface IJob
 {
 
-	public function isDue(DateTime $dateTime): bool;
+	public function isDue(DateTime $dateTime, ?DateTimeInterface $lastCheck = null): bool;
```

Comment:
I've thought it back-to-back and I like the idea, but I don't like current solution.

Let's introduce BC break and create `JobContext` object to `$job->isDue($context)` with `$run` (DateTime) and `$lastRun` (DateTime|NULL) and `$options` (array).

Does it make sense to you?

Reply:
Hi @f3l1x,

1) it is quite a shame that I've already rewritten it as my [fork](https://gitlab.com/elektro-potkan/php-scheduler),
   so my interests in this package is somewhat lower now...
   I have planned to close this PR and probably submit just a smaller ones I've came accross during the rewrite, like:
   
   1) Usage of plain job `$id` in the `LockingScheduler` for lock file-name - as the `$id` might contain nearly any string,
      it should use an md5 hash or something like that, I think.
   
   2) Deduplicating code of `run` method between the schedulers.

2) Back to this PR - Your idea seems really pretty to me, I did not think about it like that.
   Just to make sure, You mean something like this:
   ```php
   interface IJobContext
   {
   
   	/**
   	 * Returns the current time the job's isDue method is called
   	 */
   	public function getRun(): DateTimeInterface;
   
   	/**
   	 * Returns the last time the job's run method was called
   	 * @return DateTimeInterface|null - time of last run or
   	 *   null if the job did not run before or
   	 *        if the calling scheduler does not support this feature
   	 */
   	public function getLastRun(): ?DateTimeInterface;
   
   	/**
   	 * @return mixed[]
   	 */
   	public function getOptions(): array;
   
   }
   ```
   
   1) I would go for `DateTimeInterface` instead of `DateTime`.
      From my point of view - it is the scheduler's implementation detail,
      which one it would pass to the job.
      Moreover, the schedulers should be passing `DateTimeImmutable`,
      as passing the `DateTime` instance without cloning it first allows the job to simply
      modify it by mistake during the checks...
      The job should explicitly create its own mutable instance if it needs to modify it as part of the checks.
   
   2) I do not like the `run` name in the `JobContext`.
      Most of the time, it would be better to name it `now` as it will be called with the current time,
      but it is not optimal either.
      Probably naming it somewhat neutrally as `time` or so might be better.
      However, Your package, Your rules - if the `run` seems as good name to You, lets name it like that.
   
   3) What about the `options`?
      1) Should they be a scheduler-wide ones? Or per-job ones? Or combined?
      2) Where do we get them?
         1) Should we introduce some methods in the scheduler to set them?
         2) Or let's just keep them in the context for future use
            (it is a really good idea to have them here, no doubt about them from my side)
            and use an empty array for now in the schedulers?
   
   4) If I understand You well and You really mean the `lastRun` as a per-job information
      (instead of my previous scheduler-wide `lastCheck` suggestion), how to store it?
      During rewritting my fork, I implemented such feature as it seemed as very simple in the end,
      but I am not satisfied with storing of the timestamps - the names of the timestamp files to be precise.
      As the `IScheduler` already supports both - `add` and `remove` methods,
      it is stating a bit, that the scheduler is able to cope with quite a dynamic use-cases,
      where the jobs might be added and removed in a different order for each script run.
      But here comes the problem - we need to store per-job last-run timestamp,
      but the job's `$id` (or `$key` as its named in `add/get/remove` methods) might be different for each scheduler's run.
      No problem if the job gets some new `$id` but it will be a problem
      if we supply it with a last-run timestamp belonging to some other job.
      So far, I am using something like `md5(strval($id) . get_class($job))` in my fork, but it is not much better.

Thank You for Your reply.

Best regards

Elektro-potkan